datalad_next/tree.py

```diff
@@ -0,0 +1,793 @@
+            If paired with [CMD: --dataset-depth CMD][PY: dataset_depth PY],
+            refers to the maximum directory level to display underneath each
+            dataset.""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        dataset_depth=Parameter(
+            args=("--dataset-depth",),
+            doc="""maximum level of nested subdatasets to display""",
+            constraints=EnsureInt() & EnsureRange(min=0) | EnsureNone()),
+        include_files=Parameter(
+            args=("--include-files",),
+            doc="""include files in output display""",
+            action='store_true'),
+        include_hidden=Parameter(
+            args=("--include-hidden",),
+            doc="""include hidden files/directories in output display""",
+            action='store_true'),
+    )
+
+    _examples_ = [
+        dict(text="Display up to 3 levels of the current directory's "
+                  "subdirectories and their contents",
+             code_py="tree(depth=3, include_files=True)",
+             code_cmd="datalad tree --depth 3 --include-files"),
+        dict(text="Display all first- and second-level subdatasets of "
+                  "datasets located anywhere under /tmp (including in hidden "
+                  "directories) regardless of directory depth",
+             code_py="tree('/tmp', dataset_depth=2, include_hidden=True)",
+             code_cmd="datalad tree /tmp --dataset-depth 2 --include-hidden"),
+        dict(text="Display first- and second-level subdatasets and their "
+                  "contents, up to 1 directory deep within each dataset",
+             code_py="tree(dataset_depth=2, depth=1)",
+             code_cmd="datalad tree --dataset-depth 2 --depth 1"),
+    ]
+
+    @staticmethod
+    @datasetmethod(name='tree')
+    @eval_results
+    def __call__(
+            path='.',
+            *,
+            depth=None,
+            dataset_depth=None,
+            include_files=False,
+            include_hidden=False):
+
+        if dataset_depth is not None:
+            tree = DatasetTree(
+                Path(path),
+                max_depth=depth,
+                max_dataset_depth=dataset_depth,
+                exclude_node_func=build_excluded_node_func(
+                    include_hidden=include_hidden, include_files=include_files)
+            )
+        else:
+            tree = Tree(
+                Path(path),
+                max_depth=depth,
+                exclude_node_func=build_excluded_node_func(
+                    include_hidden=include_hidden, include_files=include_files)
+            )
+
+        for line in tree.print_line():
+            # print one line at a time to improve UX / perceived speed
+            print(line)
+        print("\n" + tree.stats() + "\n")
+
+        # return a generic OK status
+        yield get_status_dict(
+            action='tree',
+            status='ok',
+            path=path,
+        )
```

Comment:
Yes, good point! I had a feeling that the result set could be used more effectively (as opposed to an impoverished dict with just an "ok" status and the `path` attribute).

Since we want to print the tree nodes as soon as they are computed, it makes sense to use the node as "reporting unit". The result dict can yield from the `generate_nodes()` function, using the node attributes as custom dict keys. I will try to create a custom results renderer for it.

Reply:
I added a custom renderer (one result for each tree node) and a custom summary renderer, which includes both the tree 'report line' and the generic renderer (just to show the 'ok' status -- there is no error status that needs reporting, right?). It works with `-f` 'tailored' (the default tree display), 'generic', 'json[_pp]'. I only added the most relevant custom attributes in the results dict -- let me know if I should include fewer/more.

Reply:
Thanks for making this change. It moves the implementation closer to the rest of the family.

Looking at the result records for a simple example (one superdataset with one immediate subdataset, and one immediate subdirectory) I see this:

```sh
% datalad -f json_pp tree
{
  "action": "tree",
  "depth": 0,
  "node_str": "[\u001b[1;37mDS~0\u001b[0m] \u001b[1;35m/tmp/mike\u001b[0m",
  "path": "/tmp/mike",
  "status": "ok",
  "tree_stats": "0 datasets, 0 directories, 0 files",
  "type": "dataset"
}
{
  "action": "tree",
  "depth": 1,
  "node_str": "\u251c\u2500\u2500 [\u001b[1;37mDS~1\u001b[0m] \u001b[1;35msub\u001b[0m/",
  "path": "/tmp/mike/sub",
  "status": "ok",
  "tree_stats": "1 dataset, 0 directories, 0 files",
  "type": "dataset"
}
{
  "action": "tree",
  "depth": 1,
  "node_str": "\u2514\u2500\u2500 \u001b[1;34msubdir\u001b[0m/",
  "path": "/tmp/mike/subdir",
  "status": "ok",
  "tree_stats": "1 dataset, 1 directory, 0 files",
  "type": "directory"
}
```

which is rendered to 

```sh
% datalad tree
[DS~0] /tmp/mike
├── [DS~1] sub/
└── subdir/

1 dataset, 1 directory, 0 files

tree(ok): /tmp/mike (dataset)
```

I think it would make sense to exercise the split of rendering and structured reporting all the way to the end:

The `node_str` properties already seems to be redundant with the result of the information in the record. The formating (adding the ASCII coloring, specific layout decision) could move entirely into the renderer. This would also improve the speed for consumers of the structured information, because the custom renders are not even called for them -- not that it is a lot of processing that is taking place here, but it would be cleaner.

The `tree_stats` property is pre-formated for console rendering, but it would be good to also have it be three properties with the actual integer values -- again to facilitate consuming this info in external tools.

These changes would move the implementation closer to addressing the usecase of walking and annotating a directory tree. I can imagine a a number of consumers for such a tool -- including, but not limited to a future DataLad-aware GUI!


Lastly, I noticed that the stats include `0 files` unless I run the command with `--include-files`. I think it would be better (re principle-of-least-surprise) to not report on number of files when it is unknown, rather than suggesting that there are no files.

Reply:
Oh, I forgot a minor point: I think it is unnecessary to have the generic renderer produce the line

```
tree(ok): <path> (<type>)
```

If people want that, they can run `datalad -f generic tree`

Reply:
> These changes would move the implementation closer to addressing the usecase of walking and annotating a directory tree. I can imagine a a number of consumers for such a tool -- including, but not limited to a future DataLad-aware GUI!

That sounds like a great idea and something lots of currently-on-the-fence potential users would welcome, I'm sure -- just like git-GUIs (both desktop clients and Git[Hub,Lab] and the like) have lowered the barrier of entry for git.

The initial design was guided by the assumption that the tree display is all there is to the command. But you are right that this need not be the case. And if we offload that to the custom renderer, there is a much cleaner separation between node generation and output formatting, which improves the code in any case.

The results dict then becomes the crucial interface, so it needs to contain all the information that may be needed by consumers. A few questions on this:

https://github.com/datalad/datalad-next/blob/9d074fd0dcda12096e767d1f8f25e3d80bcdf142/datalad_next/tree.py#L157-L175

- Are there restrictions/conventions on the data types of the dict values? I tried to include only built-in types that are convertible to JSON (for the JSON renderers), such as `path` as string instead of a `Path` object.
- What about nested dicts (like the `count` attribute)? Based on a brief check through the other datalad commands, the general rule seems to be to keep the dict flat and minimal. In this case though, we need to pass all the information needed for the custom renderer (basically 'serializing' the `Tree` and `_TreeNode` (sub)classes), which adds up a number of attributes. Do you think there is a more appropriate pattern here that we could use?
- Is it okay to include an attribute conditionally, like the dataset-related attributes only for dataset nodes? Or should there be a fixed set of attributes, with value `None` where not applicable? (Again, this relates to the expectations on the side of the consumers).

Reply:
> Lastly, I noticed that the stats include 0 files unless I run the command with --include-files. I think it would be better (re principle-of-least-surprise) to not report on number of files when it is unknown, rather than suggesting that there are no files.

Done. Indeed, that is in line with the behaviour of `tree` when using the `-d` switch for directories-only.

Reply:
> Oh, I forgot a minor point: I think it is unnecessary to have the generic renderer produce the line
> 
> ```
> tree(ok): <path> (<type>)
> ```
> 
> If people want that, they can run `datalad -f generic tree`

Agree. Done.